applications/nrf_desktop/src/modules/cpi_state.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <misc/util.h>
+#include <misc/byteorder.h>
+
+#include "motion_sensor.h"
+
+#include "button_event.h"
+#include "config_event.h"
+#include "power_event.h"
+
+#define MODULE cpi_state
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_CPI_STATE_LOG_LEVEL);
+
+#define KEY_ID_INCREMENT CONFIG_DESKTOP_CPI_STATE_TOGGLE_UP_BUTTON_ID
+#define KEY_ID_DECREMENT CONFIG_DESKTOP_CPI_STATE_TOGGLE_DOWN_BUTTON_ID
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_OFF,
+};
+
+static enum state state;
+static u32_t cpi = CONFIG_DESKTOP_MOTION_SENSOR_CPI;
+static int step_cnt;
+static struct config_event *sent_config_event;
+static bool cpi_update_required;
+
+
+static void send_cpi_config_event(u32_t cpi_send)
+{
+	__ASSERT_NO_MSG(sent_config_event == NULL);
+
+	struct config_event *event = new_config_event(sizeof(cpi_send));
+
+	event->id = SETUP_EVENT_ID(SETUP_MODULE_SENSOR, SENSOR_OPT_CPI);
+	sys_put_le32(cpi_send, event->dyndata.data);
+	event->store_needed = true;
+
+	EVENT_SUBMIT(event);
+
+	sent_config_event = event;
+}
+
+static void update_cpi(void)
+{
+	BUILD_ASSERT(MOTION_SENSOR_CPI_MIN > 0);
+	BUILD_ASSERT(MOTION_SENSOR_CPI_MAX > MOTION_SENSOR_CPI_MIN);
+
+	if ((state == STATE_OFF) || (sent_config_event != NULL)) {
+		return;
+	}
+
+	int update = step_cnt * CONFIG_DESKTOP_CPI_STATE_STEP_CHANGE;
+	int new_cpi = (int)cpi + update;
+
+	if (new_cpi < MOTION_SENSOR_CPI_MIN) {
+		new_cpi = MOTION_SENSOR_CPI_MIN;
+	} else if (new_cpi > MOTION_SENSOR_CPI_MAX) {
+		new_cpi = MOTION_SENSOR_CPI_MAX;
+	}
+
+	if (((u32_t) new_cpi != cpi) || cpi_update_required) {
+		cpi = (u32_t) new_cpi;
+		send_cpi_config_event(cpi);
+	}
+
+	cpi_update_required = false;
+	step_cnt = 0;
+}
+
+static void power_down(void)
+{
+	state = STATE_OFF;
+	module_set_state(MODULE_STATE_OFF);
+}
+
+static void wake_up(void)
+{
+	state = STATE_ACTIVE;
+	module_set_state(MODULE_STATE_READY);
+	update_cpi();
+}
+
+static bool is_my_config_id(u8_t config_id)
+{
+	return (GROUP_FIELD_GET(config_id) == EVENT_GROUP_SETUP) &&
+	       (MOD_FIELD_GET(config_id) == SETUP_MODULE_SENSOR) &&
+	       (OPT_FIELD_GET(config_id) == SENSOR_OPT_CPI);
+}
+
+static bool handle_button_event(const struct button_event *event)
+{
+	if (event->pressed) {
+		if (unlikely(event->key_id == KEY_ID_INCREMENT)) {
+			step_cnt++;
+			update_cpi();
+		} else if (unlikely(event->key_id == KEY_ID_DECREMENT)) {
+			step_cnt--;
+			update_cpi();
+		}
+	}
+
+	return false;
+}
+
+static bool handle_config_event(const struct config_event *event)
+{
+	if (is_my_config_id(event->id)) {
```

Comment:
I am actually not in favor of this as we will have logic for parsing motion sensor configuration in two places...

Reply:
maybe we should keep entire logic related to this parsing here and pass a CPI value as an additional event to the sensor?

Reply:
Then we would have to assume that every mouse with configurable CPI would have this module enabled, which currently is not true.